File: packages/provider-codex/src/quota.ts
import { randomUUID } from 'node:crypto'
import {
  clampUsedPercent,
  numberHeader,
  createProviderQuota,
  severityFromUsedPercent,
  unixSecondsToIso,
  type ProviderQuota,
  type ProviderQuotaProbeResult,
  type ProviderQuotaSnapshots,
  type ProviderQuotaWindow,
} from '@demicodes/provider'
import { FileCodexAuthStore, type CodexAuthStore } from './auth'
import { z } from 'zod'
import { buildCodexHeaders } from './provider'

export interface CodexQuotaOptions {
  providerId?: string
  codexHome?: string
  baseUrl?: string
  authStore?: CodexAuthStore
  fetch?: (
    input: string | URL | Request,
    init?: RequestInit
  ) => Promise<Response>
  userAgent?: string
  /** Where the latest snapshot is kept across rebuilds and restarts (`createProviderQuota`). */
  snapshots?: ProviderQuotaSnapshots
}

const DEFAULT_CHATGPT_BASE_URL = 'https://chatgpt.com/backend-api'

/** The account's usage status, as the ChatGPT backend answers `GET /wham/usage`. */
const usageWindowSchema = z.object({
  used_percent: z.number(),
  limit_window_seconds: z.number(),
  reset_at: z.number(),
}).loose()
const usageStatusSchema = z.object({
  plan_type: z.string().optional(),
  rate_limit: z.object({
    primary_window: usageWindowSchema.nullish(),
    secondary_window: usageWindowSchema.nullish(),
  }).loose().nullish(),
}).loose()

/**
 * Codex consumer rate windows, from two places that agree:
 * - probe(): the account's usage status, `GET …/wham/usage`, which the
 *   open-source Codex CLI reads for its own status. It spends nothing and
 *   answers while the limit is reached, when a request would be refused.
 * - observeResponse: `x-codex-{primary,secondary}-used-percent` /
 *   `-window-minutes` / `-reset-at` on any live Responses call.
 * The usage status belongs to a ChatGPT sign-in; an API key has none.
 */
export function createCodexQuota(options: CodexQuotaOptions = {}): ProviderQuota {
  const providerId = options.providerId ?? 'codex'
  const authStore = options.authStore ?? new FileCodexAuthStore({
    codexHome: options.codexHome
  })
  const fetchImpl = options.fetch ?? fetch

  return createProviderQuota({
    providerId,
    snapshots: options.snapshots,
    canProbe: true,
    canObserve: true,
    probeCost: 'free',
    staleAfterMs: 5 * 60_000,
    probe: async ({ signal } = {}) => {
      const auth = await authStore.resolveAuth()
      if (auth.kind === 'apiKey')
        throw new Error('An OpenAI API key has no Codex usage status; its windows come from responses')
      const status = await authStore.status()
      const accountLabel = status.status === 'authenticated'
        ? status.accountLabel ?? null
        : null
      const headers = buildCodexHeaders(
        auth,
        { sessionId: 'demi-codex-quota-probe', requestId: randomUUID() },
        { userAgent: options.userAgent },
      )
      const base = (options.baseUrl ?? DEFAULT_CHATGPT_BASE_URL).replace(/\/+$/, '')
      const response = await fetchImpl(`${base}/wham/usage`, { method: 'GET', headers, signal })
      if (!response.ok) {
        await response.body?.cancel().catch(() => {})
        throw new Error(`Codex usage request failed (HTTP ${response.status})`)
      }
      const usage = usageStatusSchema.parse(await response.json())
      return {
        ...mapCodexUsageStatus(usage),
        accountLabel,
        raw: usage,
      }
    },
    observe: ({ headers }) => mapCodexRateLimitHeaders(headers),
  })
}

/** The plan and the two windows of a usage status. */
export function mapCodexUsageStatus(usage: z.infer<typeof usageStatusSchema>): ProviderQuotaProbeResult {
  const windows = (['primary', 'secondary'] as const).flatMap((kind) => {
    const window = usage.rate_limit?.[`${kind}_window`]
    if (!window)
      return []
    const usedPercent = clampUsedPercent(window.used_percent)
    return [{
      id: kind,
      label: windowLabel(kind, window.limit_window_seconds / 60),
      usedPercent,
      unit: 'percent' as const,
      resetsAt: unixSecondsToIso(window.reset_at),
      severity: severityFromUsedPercent(usedPercent),
    }]
  })
  const plan = usage.plan_type
  return {
    windows,
    plan: plan ? { id: plan, label: planLabel(plan), raw: plan } : null,
  }
}

/** `plus` reads Plus; `self_serve_business_usage_based` reads Self serve business usage based. */
function planLabel(plan: string): string {
  const words = plan.replaceAll('_', ' ')
  return words.charAt(0).toUpperCase() + words.slice(1)
}

/** A window is named by its length, which is what the user knows it by; the vendor's own word for it otherwise. */
function windowLabel(kind: 'primary' | 'secondary', minutes: number | null): string {
  const name = kind === 'primary' ? 'Primary' : 'Secondary'
  if (minutes === null || !(minutes > 0))
    return name
  if (minutes % (7 * 24 * 60) === 0)
    return minutes === 7 * 24 * 60 ? 'Weekly' : `${minutes / (7 * 24 * 60)}-week`
  if (minutes % (24 * 60) === 0)
    return minutes === 24 * 60 ? 'Daily' : `${minutes / (24 * 60)}-day`
  if (minutes % 60 === 0)
    return `${minutes / 60}-hour`
  return `${minutes}-minute`
}

export function mapCodexRateLimitHeaders(
  headers: Headers | undefined
): ProviderQuotaProbeResult | null {
  if (!headers)
    return null
  const primary = parseCodexWindow(headers, 'primary')
  const secondary = parseCodexWindow(headers, 'secondary')
  const windows = [primary, secondary].filter((w): w is ProviderQuotaWindow => w !== null)
  if (windows.length === 0)
    return null
  return {
    windows,
    raw: {
      primary: headerBag(headers, 'primary'),
      secondary: headerBag(headers, 'secondary'),
    },
  }
}

function parseCodexWindow(
  headers: Headers,
  kind: 'primary' | 'secondary'
): ProviderQuotaWindow | null {
  const usedPercent = clampUsedPercent(numberHeader(
    headers,
    `x-codex-${kind}-used-percent`
  ))
  if (usedPercent == null && !headers.has(`x-codex-${kind}-used-percent`))
    return null
  const windowMinutes = numberHeader(headers, `x-codex-${kind}-window-minutes`)
  const resetAt = unixSecondsToIso(numberHeader(
    headers,
    `x-codex-${kind}-reset-at`
  ))
  // The same names the usage status gives, so a window keeps its name whichever source saw it last.
  const label = windowLabel(kind, windowMinutes)
  return {
    id: kind,
    label,
    usedPercent,
    unit: 'percent',
    resetsAt: resetAt,
    severity: severityFromUsedPercent(usedPercent),
  }
}

function headerBag(
  headers: Headers,
  kind: 'primary' | 'secondary'
): Record<string, string | null> {
  return {
    usedPercent: headers.get(`x-codex-${kind}-used-percent`),
    windowMinutes: headers.get(`x-codex-${kind}-window-minutes`),
    resetAt: headers.get(`x-codex-${kind}-reset-at`),
  }
}

File: crates/provider-codex/src/quota.rs
//! The quota of a Codex account (`usage-and-quota.md` § Codex): a free probe
//! of the account's usage status, which the Codex CLI reads for its own
//! status and which answers while the limit is reached, and the
//! `x-codex-*` headers on every response of the service.

use std::sync::Arc;

use demi_core::{QuotaPlan, QuotaUnit, QuotaWindow};
use demi_provider::quota::{
    Observation, ProbeCost, ProbeReading, QuotaError, QuotaSource, clamp_used_percent,
    header_number, severity, unix_seconds,
};
use futures_util::future::BoxFuture;
use http::{HeaderMap, HeaderValue, header::ACCEPT};
use reqwest::Url;
use serde::Deserialize;

use crate::{auth::CodexAuth, request::account_headers};

/// The two windows Codex reports.
const WINDOWS: [Kind; 2] = [Kind::Primary, Kind::Secondary];

#[derive(Clone, Copy)]
enum Kind {
    Primary,
    Secondary,
}

impl Kind {
    fn id(self) -> &'static str {
        match self {
            Self::Primary => "primary",
            Self::Secondary => "secondary",
        }
    }

    fn name(self) -> &'static str {
        match self {
            Self::Primary => "Primary",
            Self::Secondary => "Secondary",
        }
    }
}

/// The quota source of one Codex account.
pub(crate) struct CodexQuota {
    pub(crate) auth: Arc<CodexAuth>,
    pub(crate) http: reqwest::Client,
    pub(crate) usage_url: Url,
    pub(crate) user_agent: HeaderValue,
}

impl QuotaSource for CodexQuota {
    fn probe_cost(&self) -> Option<ProbeCost> {
        Some(ProbeCost::Free)
    }

    fn probe(&self) -> BoxFuture<'_, Result<ProbeReading, QuotaError>> {
        Box::pin(async move {
            let credentials = self
                .auth
                .credentials(&self.http, None)
                .await
                .map_err(|failure| failure.quota_error())?;
            let stored = self
                .auth
                .stored()
                .await
                .map_err(|failure| failure.quota_error())?;
            let mut headers = account_headers(&credentials, &self.user_agent);
            headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
            let response = self
                .http
                .get(self.usage_url.clone())
                .headers(headers)
                .send()
                .await
                .map_err(|error| {
                    QuotaError::Unavailable(format!(
                        "Codex usage request failed: {}",
                        error.without_url()
                    ))
                })?;
            if !response.status().is_success() {
                let status = response.status().as_u16();
                return Err(QuotaError::Unavailable(format!(
                    "Codex usage request failed with HTTP {status}"
                )));
            }
            let text = response.text().await.map_err(|error| {
                QuotaError::Unavailable(format!(
                    "Codex usage request failed: {}",
                    error.without_url()
                ))
            })?;
            let usage: UsageStatus =
                demi_provider::wire::decode_untagged(&text).map_err(|error| {
                    QuotaError::Invalid(format!("Codex usage status cannot be read: {error}"))
                })?;
            Ok(ProbeReading {
                plan: usage
                    .plan_type
                    .filter(|plan| !plan.is_empty())
                    .map(|plan| QuotaPlan {
                        label: plan_label(&plan),
                        id: plan,
                    }),
                account_label: Some(stored.label().label),
                windows: usage_windows(usage.rate_limit),
            })
        })
    }

    fn observe(&self, observation: Observation<'_>) -> Option<Vec<QuotaWindow>> {
        let Observation::Response { headers, .. } = observation else {
            return None;
        };
        let windows: Vec<QuotaWindow> = WINDOWS
            .into_iter()
            .filter_map(|kind| header_window(headers, kind))
            .collect();
        (!windows.is_empty()).then_some(windows)
    }
}

/// `GET /wham/usage`, as far as Demi reads it.
#[derive(Deserialize)]
struct UsageStatus {
    #[serde(default)]
    plan_type: Option<String>,
    #[serde(default)]
    rate_limit: Option<RateLimit>,
}

#[derive(Deserialize)]
struct RateLimit {
    #[serde(default)]
    primary_window: Option<UsageWindow>,
    #[serde(default)]
    secondary_window: Option<UsageWindow>,
}

/// One window of the usage status: `limit_window_seconds` in seconds and
/// `reset_at` in Unix seconds.
#[derive(Deserialize)]
struct UsageWindow {
    used_percent: f64,
    limit_window_seconds: f64,
    reset_at: f64,
}

fn usage_windows(rate_limit: Option<RateLimit>) -> Vec<QuotaWindow> {
    let Some(rate_limit) = rate_limit else {
        return Vec::new();
    };
    let windows = [
        (Kind::Primary, rate_limit.primary_window),
        (Kind::Secondary, rate_limit.secondary_window),
    ];
    windows
        .into_iter()
        .filter_map(|(kind, window)| {
            let window = window?;
            Some(window_of(
                kind,
                clamp_used_percent(window.used_percent),
                Some(window.limit_window_seconds / 60.0),
                unix_seconds(window.reset_at),
            ))
        })
        .collect()
}

/// The window the `x-codex-<kind>-*` headers report: one is present when its
/// `used-percent` header is, even with a value Demi cannot read.
fn header_window(headers: &HeaderMap, kind: Kind) -> Option<QuotaWindow> {
    let prefix = format!("x-codex-{}", kind.id());
    let used_name = format!("{prefix}-used-percent");
    headers.get(&used_name)?;
    let used_percent = header_number(headers, &used_name).and_then(clamp_used_percent);
    let minutes = header_number(headers, &format!("{prefix}-window-minutes"));
    let resets_at = header_number(headers, &format!("{prefix}-reset-at")).and_then(unix_seconds);
    Some(window_of(kind, used_percent, minutes, resets_at))
}

fn window_of(
    kind: Kind,
    used_percent: Option<f64>,
    minutes: Option<f64>,
    resets_at: Option<demi_core::Timestamp>,
) -> QuotaWindow {
    QuotaWindow {
        id: kind.id().into(),
        label: window_label(kind, minutes),
        used_percent,
        used: None,
        limit: None,
        unit: Some(QuotaUnit::Percent),
        resets_at,
        severity: severity(used_percent),
        scope: None,
    }
}

/// A window is named by its length, which is what the user knows it by, the
/// same whichever path saw it: `Weekly` and `Daily`, else the length in the
/// largest unit that divides it, else the vendor's word for it.
fn window_label(kind: Kind, minutes: Option<f64>) -> String {
    const DAY: f64 = 24.0 * 60.0;
    const WEEK: f64 = 7.0 * DAY;
    let Some(minutes) = minutes.filter(|minutes| *minutes > 0.0) else {
        return kind.name().into();
    };
    if minutes % WEEK == 0.0 {
        return if minutes == WEEK {
            "Weekly".into()
        } else {
            format!("{}-week", minutes / WEEK)
        };
    }
    if minutes % DAY == 0.0 {
        return if minutes == DAY {
            "Daily".into()
        } else {
            format!("{}-day", minutes / DAY)
        };
    }
    if minutes % 60.0 == 0.0 {
        return format!("{}-hour", minutes / 60.0);
    }
    format!("{minutes}-minute")
}

/// `plus` reads `Plus`; `self_serve_business_usage_based` reads `Self serve
/// business usage based`.
fn plan_label(plan: &str) -> String {
    let words = plan.replace('_', " ");
    let mut characters = words.chars();
    match characters.next() {
        Some(first) => first.to_uppercase().chain(characters).collect(),
        None => words,
    }
}
